File: packages/native/bluetooth/ios/BluetoothTypes.swift
import CoreBluetooth
import ExpoModulesCore

/// 蓝牙状态。`rawValue` 是发给 JS 的字符串，必须和 `src/Bluetooth.types.ts` 的
/// `BluetoothState` 联合逐个对上，也和 Android 的 `BluetoothState` 枚举保持一致。
enum BluetoothState: String {
  case on
  case off
  case resetting
  case turningOn
  case turningOff
  case unauthorized
  case unknown
  case unsupported

  /// `turningOn` / `turningOff` 是 Android 独有的中间态，iOS 永远不会返回。
  static func from(_ state: CBManagerState) -> BluetoothState {
    switch state {
    case .poweredOn: return .on
    case .poweredOff: return .off
    case .resetting: return .resetting
    case .unauthorized: return .unauthorized
    case .unsupported: return .unsupported
    case .unknown: return .unknown
    @unknown default: return .unknown
    }
  }
}

/// 把 `CBManager.authorization` 翻成 JS 侧的 `BluetoothPermission`。
enum BluetoothPermissions {
  /// iOS 的蓝牙隐私弹窗**一辈子只弹一次**，用户做过选择之后系统不会再弹，
  /// 所以除 `notDetermined` 外 `canAskAgain` 一律是 `false`——
  /// 再调 `requestPermissionAsync` 也只是原样返回，得引导用户去设置页。
  static func describe(_ authorization: CBManagerAuthorization) -> [String: Any] {
    switch authorization {
    case .allowedAlways:
      return ["status": "granted", "granted": true, "canAskAgain": false]
    case .notDetermined:
      return ["status": "undetermined", "granted": false, "canAskAgain": true]
    case .denied:
      return ["status": "denied", "granted": false, "canAskAgain": false]
    case .restricted:
      // 家长控制 / MDM 限制，用户自己在设置里也改不了
      return ["status": "denied", "granted": false, "canAskAgain": false]
    @unknown default:
      return ["status": "undetermined", "granted": false, "canAskAgain": true]
    }
  }
}

/// `BluetoothResult` 的原生构造。
///
/// 字段和 TS 侧的判别联合一一对应：成功分支 code / message 为 null，失败分支 payload 为 null。
/// **两个分支的键集合必须完全一致**，缺的那些要显式给 `NSNull()` 而不是不写，
/// 否则 JS 侧拿到的是 `undefined`，判别联合就不闭合了。
enum BluetoothResults {
  static func ok(_ payload: [String: Any]) -> [String: Any] {
    ["ok": true, "code": NSNull(), "message": NSNull(), "payload": payload]
  }

  static func fail(_ code: String, _ message: String) -> [String: Any] {
    ["ok": false, "code": code, "message": message, "payload": NSNull()]
  }
}

File: packages/native/bluetooth/android/src/main/java/expo/modules/bluetooth/BluetoothTypes.kt
package expo.modules.bluetooth

import android.os.Bundle
import androidx.core.os.bundleOf

/**
 * 蓝牙状态。`value` 是发给 JS 的字符串，必须和 `src/Bluetooth.types.ts` 的
 * `BluetoothState` 联合逐个对上。
 *
 * 这里刻意不用 `BluetoothAdapter` 的 int 常量原样透传：
 * iOS 的 `CBManagerState` 取值完全不同，两端只有统一成字符串才能给 JS 一份类型。
 */
enum class BluetoothState(val value: String) {
  ON("on"),
  OFF("off"),
  TURNING_ON("turningOn"),
  TURNING_OFF("turningOff"),

  /** 取不到适配器：模拟器或确实没有蓝牙硬件 */
  UNSUPPORTED("unsupported"),

  /** 缺 BLUETOOTH_CONNECT，读状态被系统拒绝 */
  UNAUTHORIZED("unauthorized"),

  /** 适配器给了个我们不认识的取值，当作不可用处理 */
  UNKNOWN("unknown")

  // RESETTING 是 iOS 独有的（系统在重启蓝牙栈），Android 没有对应状态
}

/**
 * `BluetoothResult` 的原生构造。
 *
 * 字段和 TS 侧的判别联合一一对应：成功分支 code / message 为 null，
 * 失败分支 payload 为 null。**两个分支的键集合必须完全一致**，
 * 否则 JS 侧解构出来会是 `undefined` 而不是 `null`，判别联合就不闭合了。
 */
internal object BluetoothResults {
  fun ok(payload: Bundle?): Bundle = bundleOf(
    "ok" to true,
    "code" to null,
    "message" to null,
    "payload" to payload
  )

  fun fail(code: String, message: String): Bundle = bundleOf(
    "ok" to false,
    "code" to code,
    "message" to message,
    "payload" to null
  )
}
